### PR TITLE
Fix incorrect path for CMake components with identical framework names

### LINKED ISSUE #10140: [bug] Conan returns incorrect path for Apple Framework when multiple components have the same framework name
### Environment Details (include every applicable attribute)
  * Operating System+version: macOS 12.0.1 (21A559)
  * Compiler+version: Apple clang version 13.0.0 (clang-1300.0.29.3) - Xcode 13.1 (latest)
  * Conan version: 1.42.2
  * Python version: Python 3.9.9

### Description

Since CMake [does not yet natively support XCFrameworks](https://gitlab.kitware.com/cmake/cmake/-/issues/21752) I'm looking for alternative solutions to integrating Apple's new format into our existing codebase.
After some research, I determined that [Conan Components](https://docs.conan.io/en/latest/creating_packages/package_information.html#using-components) together with the `cmake_find_package` generator are a perfect fit for the task at hand.
I will not go in detail what XCFrameworks are, but for anyone interested, the following blog post has a lot of useful info: [PSPDFKit - Supporting XCFrameworks](https://pspdfkit.com/blog/2020/supporting-xcframeworks/).
An XCFramework is basically a folder (+ some metadata) containing different architecture builds (slices) of a framework. It has the following structure:

```
ExampleFramework.xcframework
├── Info.plist
├── ios-arm64
│   └── ExampleFramework.framework
├── ios-x86_64-simulator
│   └── ExampleFramework.framework
├── tvos-arm64
│   └── ExampleFramework.framework
├── tvos-arm64_x86_64-simulator
│   └── ExampleFramework.framework
├── watchos-arm64_32_armv7k
│   └── ExampleFramework.framework
└── watchos-arm64_x86_64-simulator
    └── ExampleFramework.framework
```

The `conanfile.py` for Example Framework looks like this (simplified):

```python
def build(self):
    # Build all slices
    for platform in self._supported_platforms:
        archive_cmd = f"xcrun xcodebuild archive -quiet -project ExampleProject.xcodeproj -scheme 'ExampleFramework' -configuration {self.settings.build_type} -archivePath '{platform}.xcarchive'"
        self.run(archive_cmd)

    # Combine all slices
    create_cmd = f"xcrun xcodebuild -create-xcframework <args> -output '{self.name}.xcframework'"
    self.run(create_cmd)

# "Component" package
def package_info(self):
    self.cpp_info.name = self.name #"ExampleFramework"

    for platform in self._supported_platforms:
        arch_dir = self._arch_for_platform[platform]
        framework_path = os.path.join(self.package_folder, f"{self.name}.xcframework", arch_dir)

        self.cpp_info.components[platform].names["cmake_find_package"] = platform
        self.cpp_info.components[platform].frameworkdirs = [framework_path]
        self.cpp_info.components[platform].frameworks = [self.name]
```

For each supported architecture of the XCFramework, a component is defined with the name of the platform, e.g: iphoneos, iphonesimulator, macosx (names used by Xcode).
__NOTE:__ The framework name specified in `self.cpp_info.components[platform].frameworks` is identical for every component (`self.name`) and this __can't__ be trivially renamed (or symlinked) to e.g. `ExampleFramework_macosx.framework`.

The test_package is a bare minimum CMake project integrating this, with the following `CMakeLists.txt`:

```cmake
cmake_minimum_required(VERSION 3.1)
project(PackageTest)

include(CMakePrintHelpers)
set(CMAKE_C_FLAGS "-fmodules")
set(CMAKE_VERBOSE_MAKEFILE ON)

# cmake_find_package generator required
find_package(ExampleFramework REQUIRED COMPONENTS macosx)

add_executable(example example.m)
target_link_libraries(example ExampleFramework::macosx)
```

Contrary to expectations, running `conan create` yields the following link error:

```
/Applications/CMake-3.21.2.app/Contents/bin/cmake -E cmake_link_script CMakeFiles/example.dir/link.txt --verbose=1
/Applications/Xcode-13.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -fmodules -O3 -DNDEBUG -arch x86_64 -isysroot /Applications/Xcode-13.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk -mmacosx-version-min=10.14 -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/example.dir/example.m.o -o example -F/Users/user/.conan/data/ExampleFramework/0.1.0/user/testing/package/<package-id>/ExampleFramework.xcframework/ios-arm64  -Wl,-rpath,/Users/user/.conan/data/ExampleFramework/0.1.0/user/testing/package/<package-id>/ExampleFramework.xcframework/ios-arm64 -framework ExampleFramework
ld: warning: ignoring file /Users/user/.conan/data/ExampleFramework/0.1.0/user/testing/package/<package-id>/ExampleFramework.xcframework/ios-arm64/ExampleFramework.framework/ExampleFramework, building for macOS-x86_64 but attempting to link with file built for iOS-arm64
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_TestClass", referenced from:
      objc-class-ref in example.m.o
ld: symbol(s) not found for architecture x86_64
```

It is very interesting to note that __although__ we specified the __macosx__ component, what `FindExampleFramework.cmake` configured was actually the path leading to the __iphoneos__ framework in the subfolder `ios-arm64`.

After a bit of debugging, by inspecting the variables defined by the custom FindExampleFramework.cmake, we observe the following:
- ✅ `ExampleFramework_macosx_FRAMEWORK_DIRS: /Users/user/.conan/data/ExampleFramework/0.1.0/user/testing/package/<package-id>/ExampleFramework.xcframework/macos-arm64_x86_64`
- ❌ `ExampleFramework_macosx_LINK_LIBS: /Users/user/.conan/data/ExampleFramework/0.1.0/user/testing/package/<package-id>/ExampleFramework.xcframework/ios-arm64/ExampleFramework.framework`
- ❌ `ExampleFramework_macosx_LIBS_FRAMEWORKS_DEPS: /Users/user/.conan/data/ExampleFramework/0.1.0/user/testing/package/<package-id>/ExampleFramework.xcframework/ios-arm64/ExampleFramework.framework`

and from this, the target property is also incorrect:
❌ `ExampleFramework::macosx.INTERFACE_LINK_LIBRARIES = "/Users/user/.conan/data/ExampleFramework/0.1.0/user/testing/package/<package-id>/ExampleFramework.xcframework/ios-arm64/ExampleFramework.framework;$<$<STREQUAL:$<TARGET_PROPERTY:TYPE>,SHARED_LIBRARY>:>;$<$<STREQUAL:$<TARGET_PROPERTY:TYPE>,MODULE_LIBRARY>:>;$<$<STREQUAL:$<TARGET_PROPERTY:TYPE>,EXECUTABLE>:>"`

### Solution

It seems that Conan's cmake_find_package generator cannot deal with different components declaring the same framework name. I have traced the behavior to the macro `conan_find_apple_frameworks` in `conans/client/generators/cmake_find_package_common.py`.
Specifically, it checks `if(CONAN_FRAMEWORK_${_FRAMEWORK}_FOUND)` but this global variable doesn't take the different `FRAMEWORKS_DIRS` given to the macro, which can be a disjoint-sets between different components.

The easiest solution to this is to add an `unset(CONAN_FRAMEWORK_${_FRAMEWORK}_FOUND CACHE)` immediately before the `find_library` call.
Personally, I think this is acceptable, as the macro throws a `FATAL_ERROR` if a framework isn't found.

From CMake's [find_library documentation](https://cmake.org/cmake/help/latest/command/find_library.html):
> If the library is found the result is stored in the variable and the search will not be repeated unless the variable is cleared.

I verified that this works as expected by monkey-patching the macro in `test_package/conanfile.py`:
```python
import textwrap
from conans.client.generators.cmake_find_package_common import CMakeFindPackageCommonMacros
CMakeFindPackageCommonMacros.apple_frameworks_macro = textwrap.dedent("""
    macro(conan_find_apple_frameworks FRAMEWORKS_FOUND FRAMEWORKS FRAMEWORKS_DIRS)
        if(APPLE)
            foreach(_FRAMEWORK ${FRAMEWORKS})
                # BUGFIX: Different components with the same framework name should still attempt to find the library
                unset(CONAN_FRAMEWORK_${_FRAMEWORK}_FOUND CACHE)

                # https://cmake.org/pipermail/cmake-developers/2017-August/030199.html
                find_library(CONAN_FRAMEWORK_${_FRAMEWORK}_FOUND NAME ${_FRAMEWORK} PATHS ${FRAMEWORKS_DIRS} CMAKE_FIND_ROOT_PATH_BOTH)
                if(CONAN_FRAMEWORK_${_FRAMEWORK}_FOUND)
                    list(APPEND ${FRAMEWORKS_FOUND} ${CONAN_FRAMEWORK_${_FRAMEWORK}_FOUND})
                else()
                    message(FATAL_ERROR "Framework library ${_FRAMEWORK} not found in paths: ${FRAMEWORKS_DIRS}")
                endif()
            endforeach()
        endif()
    endmacro()
""")
```

A more elaborate solution would require defining separate `CONAN_FRAMEWORK_` variables for each component, and e.g. passing the component name to this function as a suffix.

In any case, I cannot see a reason why this use-case shouldn't be supported (same framework name in different components) and it would certainly make the lives of all Apple devs using Conan much easier in the future.

I'll be very happy to provide more information and code if necessary, as well as answer any questions that might arise.

### PR DESCRIPTION
**Pull request by T1T4N**
Changelog: Bugfix: Incorrect path found for different components with the same framework names.
Docs: omit
Closes: https://github.com/conan-io/conan/issues/10140

This PR is for issue [#10140](https://github.com/conan-io/conan/issues/10140) – a detailed description of the problem (and the solution) can be found there.

Short summary: If multiple components in a recipe have `self.cpp_info.components[component].frameworks` defined as an identical value, Conan's `cmake_find_package` generator will incorrectly return the path to the **first** found component for **all** of them because the value is cached, even if each component points to a different path.

**NOTE**: This change is in line with many other Conan *macros* which properly clear a value before using it.

- [x] Refer to the issue that supports this Pull Request: [#10140](https://github.com/conan-io/conan/issues/10140)
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

### CODE REVIEW

**Review by memsharded — CHANGES_REQUESTED** [review]
Hi @T1T4N 

Thanks very much for your PR.
However it is targeting the legacy ``cmake_find_package`` generator. This should be checked first against ``CMakeDeps`` new one, and only in relevant cases and if risk is super low, we can backport to legacy generators (this could be the case)

A unittest would be necessary, the best thing would do:
- Do a unittest that fails, using CMakeDeps
- Propose the fix for CMakeDeps
- Once looking good, propose the same for legacy generator


This can be in this PR, I don't mean multiple PRs. But Conan 2.0 is almost there, and we are not touching legacy generators except for regressions, or zero risk bug fixes that make sense.

Thanks very much!

**Comment by T1T4N** [comment]
@memsharded Thanks for the quick reply!

I already identified that CMakeDeps is using the exact same CMake macro text, so I'll update it there as well.
I'll need some time to get acquainted with your project's unit tests, but it shouldn't be hard to have this behavior testable.
Stay tuned for new commits.

**Comment by T1T4N** [comment]
@memsharded I added a commit that addresses the issue in CMakeDeps as well.

I have trouble writing a proper test for this, as the bug exists in a function that is called on runtime by CMake.
The generated files clearly mark it as "Will be filled later", so it is not enough to simply test the generated `<name>-Target-debug.cmake` or `<name>-debug-data.cmake` files.

Can you point me to an example (integration?) test in the codebase where CMake is actually executed based on an input CMakeLists.txt file?

**Comment by xakod** [comment]
@memsharded sorry for ping. Are there any chances for pr?

**Comment by T1T4N** [comment]
Hello @memsharded,

I have created an example project to reproduce the issue: [CONAN-10140.zip](https://github.com/conan-io/conan/files/10447133/CONAN-10140.zip)

This project doesn't depend on **any macOS details**, instead it manually declares a few components. It should be much simpler and easier to understand the issue and test the fix.
It is written using the **new** Conan v2 generators (CMakeDeps, CMakeToolchain) but as previously said, it affects the older generators as well, as all of them use `conan_find_apple_frameworks` where the bug is.

<details>
<summary>Final structure of the example package</summary>

```
❯ tree ~/.conan/data/testproject-10140/0.1/jenkins/testing/package/<pkg-id>

├── ComponentA
│   ├── include
│   │   └── ExampleComponentsSameFrameworkName.h
│   └── libComponent.a
├── ComponentB
│   ├── include
│   │   └── ExampleComponentsSameFrameworkName.h
│   └── libComponent.a
├── ComponentC
│   ├── include
│   │   └── ExampleComponentsSameFrameworkName.h
│   └── libComponent.a
├── conaninfo.txt
└── conanmanifest.txt
```
</details>

To encounter the issue a simple `conan create` is enough.
The `CMakeLists.txt` of the test package will try to use the main package and it will additionally check if the paths for each component are correctly configured (they're not).


To test the bugfix, simply `export ENABLE_BUGFIX=1` before running `conan create`. The test package's `conanfile.py` will then patch `cmakedeps_macros.cmake` after it is generated in the `generate` step. Using the fixed macro results in correct values for each of the components' variables.

As I said previously, I couldn't add any tests as I'm not that familiar with your codebase, but I'm hoping that this small project will be enough to result in a proper fix.

**Comment by franramirez688** [comment]
Hi @T1T4N 

Thank you very much for taking the time to perform that test locally. For sure, we can help! Let me convert it into a Conan test and push it into your branch.

**Comment by T1T4N** [comment]
@franramirez688 Thank you very much for your reply!

Please ping me when you reproduce the issue locally and in case you need any further details.
Also, please note that I just updated the PR to reflect all the changes contained in the example project.

File: conan/tools/cmake/cmakedeps/templates/macros.py [hunk 1]
```diff
@@ -29,8 +29,10 @@ def template(self):
        macro(conan_find_apple_frameworks FRAMEWORKS_FOUND FRAMEWORKS FRAMEWORKS_DIRS)
            if(APPLE)
                foreach(_FRAMEWORK ${FRAMEWORKS})
+                   unset(CONAN_FRAMEWORK_${_FRAMEWORK}_FOUND CACHE)
+
                    # https://cmake.org/pipermail/cmake-developers/2017-August/030199.html
-                   find_library(CONAN_FRAMEWORK_${_FRAMEWORK}_FOUND NAMES ${_FRAMEWORK} PATHS ${FRAMEWORKS_DIRS} CMAKE_FIND_ROOT_PATH_BOTH)
+                   find_library(CONAN_FRAMEWORK_${_FRAMEWORK}_FOUND NAMES ${_FRAMEWORK} PATHS ${FRAMEWORKS_DIRS} NO_CMAKE_PATH CMAKE_FIND_ROOT_PATH_BOTH)
```
Comment (memsharded) [inline conan/tools/cmake/cmakedeps/templates/macros.py:36]: In the normal library find, we are doing:
```cmake
find_library(CONAN_FOUND_LIBRARY NAMES ${_LIBRARY_NAME} PATHS ${package_libdir}
                            NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)
```
That is, using the global NO_DEFAULT_PATH to exclude all of them, including NO_CMAKE_PATH. Maybe we want to do exactly the same here?
And also to switch and use NO_CMAKE_FIND_ROOT_PATH instead?

**Review by memsharded — APPROVED** [review]
Looking good, I will add @jcar87 for final feedback about the default paths that should be used while finding frameworks

File: conan/tools/cmake/cmakedeps/templates/macros.py (hunk 1 above)
Comment (T1T4N) [reply conan/tools/cmake/cmakedeps/templates/macros.py:36]: Hello @memsharded,

Thank you very much for your reply.
I've tested all the options listed under [find_library](https://cmake.org/cmake/help/latest/command/find_library.html), in their hierarchical order. This is the result that I observed:
1. NO_DEFAULT_PATH ✅
2. NO_PACKAGE_ROOT_PATH ❌
3. NO_CMAKE_PATH ✅
4. NO_CMAKE_ENVIRONMENT_PATH ❌
5. NO_SYSTEM_ENVIRONMENT_PATH ❌
6. NO_CMAKE_SYSTEM_PATH ❌
where ❌ - still broken & ✅ - behavior fixed

I also just tested `NO_CMAKE_FIND_ROOT_PATH` and _on it's own_ it doesn't make a difference.

As `NO_DEFAULT_PATH` disables everything, it works. I went for `NO_CMAKE_PATH` in the PR because I tried to find the lowest scope that fixes the problem.

To sum up: the combination of `NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH` seems to work, and I additionally tested it on a real Apple XCFramework. Let's see if it doesn't break any of your tests.

Comment (memsharded) [reply conan/tools/cmake/cmakedeps/templates/macros.py:36]: Then, it looks good, we can change it to that combination to be consistent with the other.
If you want to be a bit more conservative, I am fine with doing the minimal change in ``cmake_find_package``, but ``CMakeDeps`` can be changed to align completely with the other find_library.
Thanks!

Comment (T1T4N) [reply conan/tools/cmake/cmakedeps/templates/macros.py:36]: Hello @memsharded,

I rebased and updated my two commits with `NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH` to ensure consistency with the other functions in the codebase.
Thanks for all the feedback and I hope that the PR can now be merged!

**Comment by franramirez688** [comment]
Hi @T1T4N 

Sorry, I had to revert the flags `NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH`. There were a lot of tests broken because of that (we're mixing system frameworks and our own ones and it seems to be a problem). I'm going to open a new issue to have a look at this. Anyway, it was a separate issue. Thanks again!

**Comment by T1T4N** [comment]
Hello @franramirez688,

Thanks for your reply!
Should we now proceed forward with the original change: `NO_CMAKE_PATH CMAKE_FIND_ROOT_PATH_BOTH` ?

I see that your revert commit doesn't include it, and as that is part of the fix the new test should fail without it.

**Comment by franramirez688** [comment]
> Hello @franramirez688,
> 
> Thanks for your reply! Should we now proceed forward with the original change: `NO_CMAKE_PATH CMAKE_FIND_ROOT_PATH_BOTH` ?

Yes, please.

> I see that your revert commit doesn't include it, and as that is part of the fix the new test should fail without it.

Yeah, sorry, I did not realize it 😵‍💫

**Comment by franramirez688** [comment]
@T1T4N I just restored it! Thanks for your comment because I did not remember it 😭

**Review by franramirez688 — APPROVED** [review]

**Comment by franramirez688** [comment]
@T1T4N @memsharded 
@jcar87 and I have been looking at the issue, and it seems that the bug could be related to another thing. Let me have a look deeper to get what's going on there. I'll update you both later.

**Comment by franramirez688** [comment]
So sorry for the late response.

Let's wait for this issue https://github.com/conan-io/conan/issues/13018 to be tackled. IMHO, we have to give a better approach to this issue.

**Comment by memsharded** [comment]
I am closing this ticket as superseded by the new ``CMakeConfigDeps`` generator, now in incubating: https://docs.conan.io/2/incubating.html

The new generator has changed how frameworks are processed, but also how the CMake generated files are located.

Please give it a try, report updated tickets for any further question or issue.

If everything goes well, then it would be appreciated to get positive feedback in https://github.com/conan-io/conan/issues/18211, to get the feature out of incubating.

Thanks very much for your contribution and your feedback!
